File: UNIT-01/orchestrator/mcp.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"sync"
)

// ─── MCP Config (Claude Code compatible) ───────────────────────────────────────

type MCPConfig struct {
	MCPServers map[string]MCPServerConfig `json:"mcpServers"`
}

type MCPServerConfig struct {
	Command string            `json:"command"`
	Args    []string          `json:"args,omitempty"`
	Env     map[string]string `json:"env,omitempty"`
}

// ─── MCP Protocol Types ────────────────────────────────────────────────────────

type MCPTool struct {
	Name        string      `json:"name"`
	Description string      `json:"description"`
	InputSchema interface{} `json:"inputSchema"`
}

type MCPContentItem struct {
	Type string `json:"type"`
	Text string `json:"text,omitempty"`
}

type MCPToolResult struct {
	Content []MCPContentItem `json:"content"`
	IsError bool             `json:"isError,omitempty"`
}

// ─── JSON-RPC 2.0 over stdio with Content-Length framing ───────────────────────

type mcpRequest struct {
	JSONRPC string `json:"jsonrpc"`
	Method  string `json:"method"`
	Params  any    `json:"params,omitempty"`
	ID      int    `json:"id"`
}

type mcpResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *mcpRPCError    `json:"error,omitempty"`
	ID      int             `json:"id"`
}

type mcpRPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

func (e *mcpRPCError) Error() string {
	return fmt.Sprintf("MCP Error %d: %s", e.Code, e.Message)
}

// ─── MCP Server Instance ───────────────────────────────────────────────────────

type mcpServer struct {
	Name   string
	cmd    *exec.Cmd
	stdin  *bufio.Writer
	stdout *bufio.Scanner
	mu     sync.Mutex
	seqNo  int
}

func newMCPServer(name string, cfg MCPServerConfig) (*mcpServer, error) {
	cmd := exec.Command(cfg.Command, cfg.Args...)
	cmd.Env = os.Environ()
	for k, v := range cfg.Env {
		cmd.Env = append(cmd.Env, k+"="+v)
	}

	stdinPipe, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("stdin pipe: %w", err)
	}
	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("stdout pipe: %w", err)
	}
	cmd.Stderr = os.Stderr

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("start: %w", err)
	}

	s := &mcpServer{
		Name:   name,
		cmd:    cmd,
		stdin:  bufio.NewWriter(stdinPipe),
		stdout: bufio.NewScanner(stdoutPipe),
		seqNo:  0,
	}
	s.stdout.Buffer(make([]byte, 0, 64*1024), 10*1024*1024)
	s.stdout.Split(scanMCPFrame)

	return s, nil
}

// scanMCPFrame splits on Content-Length: N\r\n\r\n<JSON>\n
func scanMCPFrame(data []byte, atEOF bool) (advance int, token []byte, err error) {
	headerEnd := strings.Index(string(data), "\r\n\r\n")
	if headerEnd == -1 {
		if atEOF {
			return 0, data, bufio.ErrFinalToken
		}
		return 0, nil, nil
	}

	var contentLen int
	for _, line := range strings.Split(string(data[:headerEnd]), "\r\n") {
		if strings.HasPrefix(line, "Content-Length: ") {
			fmt.Sscanf(line, "Content-Length: %d", &contentLen)
		}
	}

	bodyStart := headerEnd + 4
	if len(data) < bodyStart+contentLen {
		return 0, nil, nil
	}

	return bodyStart + contentLen, data[bodyStart : bodyStart+contentLen], nil
}

func (s *mcpServer) roundTrip(method string, params any) (json.RawMessage, error) {
	s.mu.Lock()
	s.seqNo++
	id := s.seqNo
	s.mu.Unlock()

	req := mcpRequest{
		JSONRPC: "2.0",
		Method:  method,
		Params:  params,
		ID:      id,
	}

	data, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("marshal: %w", err)
	}

	msg := fmt.Sprintf("Content-Length: %d\r\n\r\n%s", len(data), data)
	if _, err := s.stdin.WriteString(msg); err != nil {
		return nil, fmt.Errorf("write: %w", err)
	}
	s.stdin.Flush()

	for s.stdout.Scan() {
		var resp mcpResponse
		if err := json.Unmarshal(s.stdout.Bytes(), &resp); err != nil {
			return nil, fmt.Errorf("unmarshal: %w", err)
		}
		if resp.ID == id {
			if resp.Error != nil {
				return nil, resp.Error
			}
			return resp.Result, nil
		}
		// Non-matching ID = server notification, discard
	}
	if err := s.stdout.Err(); err != nil {
		return nil, fmt.Errorf("read: %w", err)
	}
	return nil, fmt.Errorf("server %s closed connection", s.Name)
}

func (s *mcpServer) initialize() error {
	params := map[string]any{
		"protocolVersion": "2024-11-05",
		"capabilities":    map[string]any{},
		"clientInfo": map[string]string{
			"name":    "unit01",
			"version": "1.0",
		},
	}
	_, err := s.roundTrip("initialize", params)
	return err
}

func (s *mcpServer) listTools() ([]MCPTool, error) {
	result, err := s.roundTrip("tools/list", nil)
	if err != nil {
		return nil, err
	}
	var list struct {
		Tools []MCPTool `json:"tools"`
	}
	if err := json.Unmarshal(result, &list); err != nil {
		return nil, fmt.Errorf("unmarshal tools: %w", err)
	}
	return list.Tools, nil
}

func (s *mcpServer) callTool(name string, args map[string]any) (*MCPToolResult, error) {
	params := map[string]any{
		"name":      name,
		"arguments": args,
	}
	result, err := s.roundTrip("tools/call", params)
	if err != nil {
		return nil, err
	}
	var toolResult MCPToolResult
	if err := json.Unmarshal(result, &toolResult); err != nil {
		return nil, fmt.Errorf("unmarshal result: %w", err)
	}
	return &toolResult, nil
}

func (s *mcpServer) shutdown() {
	s.cmd.Process.Kill()
	s.cmd.Wait()
}

// ─── MCP Manager ───────────────────────────────────────────────────────────────

type MCPManager struct {
	servers map[string]*mcpServer
	tools   map[string]mcpToolRef
}

type mcpToolRef struct {
	ServerName string
	Tool       MCPTool
}

// LoadMCPConfig reads ~/.config/unit01/mcp.json and starts all MCP servers.
// Missing config file is not an error — returns an empty manager.
func LoadMCPConfig() *MCPManager {
	mgr := &MCPManager{
		servers: make(map[string]*mcpServer),
		tools:   make(map[string]mcpToolRef),
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return mgr
	}
	configPath := home + "/.config/unit01/mcp.json"

	data, err := os.ReadFile(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			fmt.Println("◆ MCP: no config at ~/.config/unit01/mcp.json, skipping")
		}
		return mgr
	}

	var cfg MCPConfig
	if err := json.Unmarshal(data, &cfg); err != nil {
		fmt.Printf("⚠️ MCP config parse error: %v\n", err)
		return mgr
	}

	for name, srvCfg := range cfg.MCPServers {
		server, err := newMCPServer(name, srvCfg)
		if err != nil {
			fmt.Printf("⚠️ MCP [%s] start failed: %v\n", name, err)
			continue
		}
		if err := server.initialize(); err != nil {
			fmt.Printf("⚠️ MCP [%s] init failed: %v\n", name, err)
			server.shutdown()
			continue
		}
		tools, err := server.listTools()
		if err != nil {
			fmt.Printf("⚠️ MCP [%s] list_tools failed: %v\n", name, err)
			server.shutdown()
			continue
		}

		mgr.servers[name] = server
		for _, tool := range tools {
			mgr.tools[tool.Name] = mcpToolRef{
				ServerName: name,
				Tool:       tool,
			}
		}
		fmt.Printf("◆ MCP [%s] loaded (%d tools)\n", name, len(tools))
	}

	return mgr
}

// ToolCount returns the number of registered MCP tools.
func (m *MCPManager) ToolCount() int {
	return len(m.tools)
}

// ToolDescriptions returns a string listing all MCP tools for prompt injection.
func (m *MCPManager) ToolDescriptions() string {
	if len(m.tools) == 0 {
		return ""
	}
	var b strings.Builder
	b.WriteString("\n\n### MCP EXTENSIBLE TOOLS (available on this system):\n")
	for _, ref := range m.tools {
		b.WriteString(fmt.Sprintf("- <%s>", ref.Tool.Name))
		if ref.Tool.Description != "" {
			b.WriteString(": " + ref.Tool.Description)
		}
		b.WriteString("\n")
	}
	b.WriteString("Use these with the same <tool_name arg=\"val\" /> syntax as the directive tags above.")
	return b.String()
}

// ExecuteMCPTool routes a tool call to the correct MCP server.
// Returns empty string if the tool is not found.
func (m *MCPManager) ExecuteMCPTool(name string, args map[string]any) string {
	ref, ok := m.tools[name]
	if !ok {
		return ""
	}
	server, ok := m.servers[ref.ServerName]
	if !ok {
		return ""
	}
	result, err := server.callTool(name, args)
	if err != nil {
		return fmt.Sprintf("❌ MCP Error (%s): %v", name, err)
	}
	var sb strings.Builder
	for _, item := range result.Content {
		sb.WriteString(item.Text)
	}
	return sb.String()
}

// HasTool checks if an MCP tool with the given name exists.
func (m *MCPManager) HasTool(name string) bool {
	_, ok := m.tools[name]
	return ok
}

// Shutdown kills all MCP server processes.
func (m *MCPManager) Shutdown() {
	for name, s := range m.servers {
		s.shutdown()
		fmt.Printf("◆ MCP [%s] stopped\n", name)
	}
}

File: UNIT-01/orchestrator/src/mcp.rs
use serde::Deserialize;
use serde_json::Value;
use std::collections::HashMap;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::{Child, Command};
use tokio::sync::Mutex;

#[derive(Debug, Clone, Deserialize)]
pub struct MCPConfig {
    #[serde(rename = "mcpServers")]
    pub mcp_servers: HashMap<String, MCPServerConfig>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MCPServerConfig {
    pub command: String,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub env: HashMap<String, String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MCPTool {
    pub name: String,
    pub description: String,
    pub input_schema: Value,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MCPToolResult {
    pub content: Vec<MCPContentItem>,
    #[serde(default)]
    pub is_error: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MCPContentItem {
    #[serde(rename = "type")]
    pub item_type: String,
    #[serde(default)]
    pub text: String,
}

struct MCPServerInstance {
    name: String,
    child: Child,
    stdin: tokio::io::BufWriter<tokio::process::ChildStdin>,
    stdout: BufReader<tokio::process::ChildStdout>,
    seq_no: u32,
}

impl MCPServerInstance {
    async fn send_request(&mut self, method: &str, params: Value) -> Result<Value, String> {
        self.seq_no += 1;
        let id = self.seq_no;

        let req = serde_json::json!({
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": id
        });

        let data = serde_json::to_vec(&req).map_err(|e| format!("serialize: {}", e))?;
        let header = format!("Content-Length: {}\r\n\r\n", data.len());
        let mut full = header.into_bytes();
        full.extend(data);

        self.stdin
            .write_all(&full)
            .await
            .map_err(|e| format!("write: {}", e))?;
        self.stdin
            .flush()
            .await
            .map_err(|e| format!("flush: {}", e))?;

        let mut line = String::new();
        loop {
            line.clear();
            match self.stdout.read_line(&mut line).await {
                Ok(0) => return Err("connection closed".to_string()),
                Ok(_) => {
                    let trimmed = line.trim();
                    if trimmed.is_empty() {
                        continue;
                    }
                    if let Ok(resp) = serde_json::from_str::<Value>(trimmed) {
                        if resp.get("id").and_then(|v| v.as_u64()) == Some(id as u64) {
                            if let Some(err) = resp.get("error") {
                                return Err(format!("MCP error: {}", err));
                            }
                            if let Some(result) = resp.get("result") {
                                return Ok(result.clone());
                            }
                        }
                    }
                }
                Err(e) => return Err(format!("read: {}", e)),
            }
        }
    }
}

pub struct MCPManager {
    servers: Mutex<HashMap<String, MCPServerInstance>>,
    tools: Mutex<HashMap<String, MCPToolRef>>,
}

#[derive(Clone)]
struct MCPToolRef {
    server_name: String,
    tool: MCPTool,
}

impl MCPManager {
    pub fn new() -> Self {
        Self {
            servers: Mutex::new(HashMap::new()),
            tools: Mutex::new(HashMap::new()),
        }
    }

    pub async fn load_config(&self) {
        let config_path = if let Some(home) = home::home_dir() {
            format!("{}/.config/unit01/mcp.json", home.display())
        } else {
            return;
        };

        let data = match tokio::fs::read_to_string(&config_path).await {
            Ok(d) => d,
            Err(_) => return,
        };

        let cfg: MCPConfig = match serde_json::from_str(&data) {
            Ok(c) => c,
            Err(e) => {
                eprintln!("⚠️ MCP config parse error: {}", e);
                return;
            }
        };

        for (name, srv_cfg) in cfg.mcp_servers {
            match self.spawn_server(&name, &srv_cfg).await {
                Ok(()) => {}
                Err(e) => eprintln!("⚠️ MCP [{}] failed: {}", name, e),
            }
        }
    }

    async fn spawn_server(&self, name: &str, cfg: &MCPServerConfig) -> Result<(), String> {
        let mut cmd = Command::new(&cfg.command);
        cmd.args(&cfg.args);
        cmd.env_clear();
        cmd.envs(std::env::vars());
        for (k, v) in &cfg.env {
            cmd.env(k, v);
        }

        cmd.stdin(std::process::Stdio::piped());
        cmd.stdout(std::process::Stdio::piped());
        cmd.stderr(std::process::Stdio::inherit());

        let mut child = cmd.spawn().map_err(|e| format!("spawn: {}", e))?;

        let stdin = child.stdin.take().ok_or("no stdin")?;
        let stdout = child.stdout.take().ok_or("no stdout")?;

        let mut instance = MCPServerInstance {
            name: name.to_string(),
            child,
            stdin: tokio::io::BufWriter::new(stdin),
            stdout: BufReader::new(stdout),
            seq_no: 0,
        };

        instance
            .send_request(
                "initialize",
                serde_json::json!({
                    "protocolVersion": "2024-11-05",
                    "capabilities": {},
                    "clientInfo": { "name": "unit01", "version": "1.0" }
                }),
            )
            .await?;

        let tools_result = instance
            .send_request("tools/list", serde_json::json!(null))
            .await?;
        let tools_list: Vec<MCPTool> = serde_json::from_value(
            tools_result
                .get("tools")
                .cloned()
                .unwrap_or(serde_json::Value::Array(vec![])),
        )
        .map_err(|e| format!("tool parse: {}", e))?;

        let mut servers = self.servers.lock().await;
        servers.insert(name.to_string(), instance);

        let mut tools = self.tools.lock().await;
        for tool in tools_list {
            tools.insert(
                tool.name.clone(),
                MCPToolRef {
                    server_name: name.to_string(),
                    tool,
                },
            );
        }

        Ok(())
    }

    pub async fn tool_count(&self) -> usize {
        self.tools.lock().await.len()
    }

    pub async fn tool_descriptions(&self) -> String {
        let tools = self.tools.lock().await;
        if tools.is_empty() {
            return String::new();
        }
        let mut desc = String::from("\n\n### MCP EXTENSIBLE TOOLS (available on this system):\n");
        for (_name, ref_) in tools.iter() {
            desc.push_str(&format!("- <{}>", ref_.tool.name));
            if !ref_.tool.description.is_empty() {
                desc.push_str(&format!(": {}", ref_.tool.description));
            }
            desc.push('\n');
        }
        desc
    }

    pub async fn execute_tool(&self, name: &str, args: Value) -> Option<String> {
        let tools = self.tools.lock().await;
        let ref_ = tools.get(name)?.clone();
        drop(tools);
        let mut servers = self.servers.lock().await;
        let instance = servers.get_mut(&ref_.server_name)?;

        match instance
            .send_request(
                "tools/call",
                serde_json::json!({
                    "name": name,
                    "arguments": args
                }),
            )
            .await
        {
            Ok(result) => {
                let tr: MCPToolResult = serde_json::from_value(result).ok()?;
                let mut text = String::new();
                for item in tr.content {
                    text.push_str(&item.text);
                }
                Some(text)
            }
            Err(_) => None,
        }
    }

    pub async fn has_tool(&self, name: &str) -> bool {
        self.tools.lock().await.contains_key(name)
    }

    pub async fn shutdown(&self) {
        let mut servers = self.servers.lock().await;
        for (_name, instance) in servers.iter_mut() {
            let _ = instance.child.kill().await;
        }
        servers.clear();
    }
}
